Behat tests for orders in the Back Office. Various Behat features created and scenarios added

tdavidsonas88 opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Behat tests for orders in the Back Office
| Type?         | improvement
| Category?     | TE
| BC breaks?    |  no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/14480
| How to test?  | QA testing not needed. Travis build should be green

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/16525)
<!-- Reviewable:end -->
Handlers coverage:
CreateEmptyCustomerHandler
UpdateCartAddressesHandler
UpdateProductQuantityInCartHandler
AddOrderFromBackOfficeHandler
UpdateOrderStatusHandler
GetOrderForViewingHandler
UpdateOrderShippingDetailsHandler
AddPaymentHandler
DuplicateOrderCartHandler
GetCartInformationHandler
SearchProductsHandler
AddProductToOrderHandler
GenerateInvoiceHandler
SetFreeShippingToCartHandler
BulkChangeOrderStatusHandler
ChangeOrderDeliveryAddressHandler

tdavidsonas88 committed:
Update order status feature behat tests

tdavidsonas88 committed:
Update order status feature behat tests

refactoring after Rokas code review

tdavidsonas88 committed:
Update order status feature behat tests

refactoring after code review

tdavidsonas88 committed:
Update order status feature behat tests

refactoring after Rokas code review

tdavidsonas88 committed:
Update order status feature behat tests

PrettyCI — Code formatting fixed

tdavidsonas88 committed:
feature/order-edit-shipping=integration-tests

tests refactored
new feature update_order_shipping_details tests added

tdavidsonas88 commented:
Travis fails because of the reason not related this pull request scope:

Tests\Integration\Behaviour\Features\Context\CLDRFeatureContext::assertDisplayPrice()
      Displayed price is "14 789,54 " but "14 789,54 €" was expected (RuntimeException)
    And database contains 1 rows of currency "EUR"

matks commented:
```
  Background:                                                # Features/Scenario/Order/update_order_shipping_details.feature:8
    Given the current currency is "USD"                      # Tests\Integration\Behaviour\Features\Context\Domain\CartFeatureContext::addCurrencyToContext()
    Given there is existing order with reference "XKBKNABJK" # Tests\Integration\Behaviour\Features\Context\Domain\OrderFeatureContext::thereIsExistingOrderWithReference()
  Scenario: Update order shipping details                                                     # Features/Scenario/Order/update_order_shipping_details.feature:12
sh: 1: -t: not found
    When I update order "XKBKNABJK" Tracking number to "TEST1234" and Carrier to "My carrier" # Tests\Integration\Behaviour\Features\Context\Domain\OrderFeatureContext::iUpdateOrderTrackingNumberToAndCarrierTo()
      An error occurred while sending an email to the customer. (PrestaShop\PrestaShop\Core\Domain\Order\Exception\TransistEmailSendingException)
    Then order "XKBKNABJK" has Tracking number "TEST1234"                                     # Tests\Integration\Behaviour\Features\Context\Domain\OrderFeatureContext::orderHasTrackingNumber()
    And order "XKBKNABJK" has Carrier "My carrier"                                            # Tests\Integration\Behaviour\Features\Context\Domain\OrderFeatureContext::orderHasCarrier()
    When I update order "XKBKNABJK" Tracking number to "TEST123" and Carrier to "0"           # Tests\Integration\Behaviour\Features\Context\Domain\OrderFeatureContext::iUpdateOrderTrackingNumberToAndCarrierTo()
    Then order "XKBKNABJK" has Tracking number "TEST123"                                      # Tests\Integration\Behaviour\Features\Context\Domain\OrderFeatureContext::orderHasTrackingNumber()
    And order "XKBKNABJK" has Carrier "0"     
```
I think we also need to disable email sending for tests. I remember seeing something to do that in `app/config/config_test.yml` (see swiftmailer config), it might need to be reused or config is false

matks commented:
> Travis fails because of the reason not related this pull request scope:
> 
> Tests\Integration\Behaviour\Features\Context\CLDRFeatureContext::assertDisplayPrice()
> Displayed price is "14 789,54 " but "14 789,54 €" was expected (RuntimeException)
> And database contains 1 rows of currency "EUR"

Well it does look like a bug 🤔 as indeed the `€` symbol is missing. @matthieu-rolland can you tell us if this looks like one of these CLDR bugs you fixed before ?

tdavidsonas88 committed:
order integration-tests with behat

Refactored all the behat tests I've created up untill now to use the Domain classes avoiding the legacy code.

started creating:
Feature: Adding products to an existing Order

tdavidsonas88 committed:
order integration-tests with behat

disabled email sending for features

tdavidsonas88 commented:
> I think we also need to disable email sending for tests. I remember seeing something to do that in `app/config/config_test.yml` (see swiftmailer config), it might need to be reused or config is false

@matks in app/config/config_test.yml swiftmailer settings are already correct:

swiftmailer:
    disable_delivery: true

Since legacy code is sending email using @Mail::Send in OrderCarrier.php, the temporary solution while code is being migrated would be changing PS_MAIL_METHOD to 3 in the testdb using:

  Background:
    Given email sending is disabled

Fixed it like this with the last commit to this branch.

tdavidsonas88 committed:
order integration-tests with behat

code style fixes after Rokas code review

tdavidsonas88 committed:
order integration-tests with behat

add_payment.feature behat tests

tdavidsonas88 committed:
order integration-tests with behat

add_payment.feature behat tests code style fixes

tdavidsonas88 committed:
add_payment.feature behat tests

testcase created for:
When I add payment to order id 1 exception is thrown with the...

tdavidsonas88 committed:
add_payment.feature behat tests

2 tests are failing because of the possible bug:

it should not be allowed to add payments with negative amount?
And if it's added then total_sum should not be negative?

tdavidsonas88 committed:
add_payment.feature behat tests

started to write logic to tests orders payments
depending on the order state

tdavidsonas88 committed:
add_payment.feature behat tests

valid invoice added for adding payment test case

tdavidsonas88 committed:
add_payment.feature behat tests

fixed the cldr.feature failures by adding missing property ->php_self
for test double AdminController std class object

Some fixes for scenarios and todo added for splitting code to
more accurate Order Page contexts

tdavidsonas88 commented:
> > Travis fails because of the reason not related this pull request scope:
> > Tests\Integration\Behaviour\Features\Context\CLDRFeatureContext::assertDisplayPrice()
> > Displayed price is "14 789,54 " but "14 789,54 €" was expected (RuntimeException)
> > And database contains 1 rows of currency "EUR"
> 
> Well it does look like a bug  as indeed the `€` symbol is missing. @matthieu-rolland can you tell us if this looks like one of these CLDR bugs you fixed before ?

@rokaszygmantas @matks actually this one is related to the test double I've created for the AdminController in the OrderFeatureContext -> @BeforeScenario -> $adminControllerTestDouble->php_self.

php_self property was missing in this case. So this failure of test is fixed.

Now the two tests are failing because of the issue created today: https://github.com/PrestaShop/PrestaShop/issues/16582

When the issue is fixed - the tests should pass;

Or If the business logic of the failing Behat tests is wrong, please let me know and I'll fix the scenarios.

tdavidsonas88 committed:
add_payment.feature behat tests

OrderPaymentFeatureContext created
related code move there

matthieu-rolland commented:
> > Travis fails because of the reason not related this pull request scope:
> > Tests\Integration\Behaviour\Features\Context\CLDRFeatureContext::assertDisplayPrice()
> > Displayed price is "14 789,54 " but "14 789,54 €" was expected (RuntimeException)
> > And database contains 1 rows of currency "EUR"
> 
> Well it does look like a bug 🤔 as indeed the `€` symbol is missing. @matthieu-rolland can you tell us if this looks like one of these CLDR bugs you fixed before ?

@matks Not really, the CLDR issues in tests were related to calculation results, not to formatting. 🤔

tdavidsonas88 committed:
add_payment.feature behat tests

OrderPaymentFeatureContext missing use statement added

tdavidsonas88 committed:
Merge branch 'm/orders/create-summary' of https://github.com/zuk3975/PrestaShop into zuk3975-m/orders/create-summary

# Conflicts:
#	admin-dev/themes/new-theme/public/order.bundle.js

margud commented on tests/Integration/Behaviour/Features/Scenario/Order/update_order_status.feature:
```diff
@@ -0,0 +1,21 @@
+    And order 2 has status "Delivered"
+
+  Scenario: Update order status
+    When I update order 1 to status "Awaiting bank wire payment"
```
maybe order with id 1?

margud commented on tests/Integration/Behaviour/Features/Context/Domain/OrderFeatureContext.php:
```diff
@@ -111,4 +181,119 @@ public function generateOrderInvoice($orderReference)
+     * @throws OrderException
+     * @throws RuntimeException
+     */
+    public function iUpdateOrdersToStatus(string $orderIdsString, string $status)
```
Update order to status not update order status?

tdavidsonas88 committed:
Refactored the naming of the methods after margud QA

tdavidsonas88 committed:
Behat Feature: Duplicate order cart from Back Office

tdavidsonas88 committed:
Behat feature test for Duplicate order cart from BO

tdavidsonas88 commented on tests/Integration/Behaviour/Features/Context/Domain/AbstractDomainFeatureContext.php:
```diff
@@ -28,18 +28,23 @@
 use Tests\Integration\Behaviour\Features\Context\CommonFeatureContext;
 use Tests\Integration\Behaviour\Features\Context\SharedStorage;
 
+require_once __DIR__ .'/../../../../../../vendor/phpunit/phpunit/src/Framework/Assert/Functions.php';
```
@rokaszygmantas @matks it would be more efficient to use phpunit assertions instead of ifs and throwing exceptions with our own efforts. Behat works nice with phpunit. Just would like to get confirmation from you guys that is a good way to go? :)

See below how I use it with:

```
        assertSame(
            $customerId,
            $customerIdReceived,
            sprintf('Expected customer with id "%s" but received "%s', $customerId, $customerIdReceived)
        );
```

tdavidsonas88 replied:
To avoid using the relative path refactored the code to use PHPUnit_Framework_Assert::assertSame

tdavidsonas88 committed:
Behat feature test for Duplicate order cart from BO

tdavidsonas88 committed:
Behat feature test for Duplicate order cart from BO

tdavidsonas88 committed:
Bussiness value defined for duplicate order cart feature

tdavidsonas88 committed:
shipping info is null by default in duplicate order cart feature

tdavidsonas88 committed:
refactored to use PHPUnit_Framework_Assert:: instead of relative path

matks commented on tests/Integration/Behaviour/Features/Context/Domain/OrderFeatureContext.php:
```diff
@@ -26,26 +26,70 @@
+        $adminControllerTestDouble = new stdClass();
+        $adminControllerTestDouble->controller_type = 'admin';
+        $adminControllerTestDouble->php_self = 'dummyTestDouble';
+        Context::getContext()->controller = $adminControllerTestDouble;
```
Waho 😭it's heart-breaking we have to use this double
But I guess we cant do it without

Nice finding though 👍

matks commented on tests/Integration/Behaviour/Features/Scenario/Order/add_product.feature:
```diff
@@ -4,9 +4,32 @@ Feature: Adding products to an existing Order
     When I add 2 products with reference "demo_5", price 16 and free shipping to order "XKBKNABJK" with new invoice
     Then order "XKBKNABJK" should contain 2 products with reference "demo_5"
+    # id_product = 4 is same as referenced by demo_5
+    When I add products with new invoice and the following properties:
```
Im sorry, I dont understand what "products with new invoice" means 😅 
Do you mean products with new price (= not the catalog price) ?

tdavidsonas88 replied:
Changed it to: `When I add products to order with new invoice with the following products properties:`
It means:  

> Add product to an order with new invoice. It applies to orders that were already paid and waiting for payment.

matks replied:
better 😃

matks commented on tests/Integration/Behaviour/Features/Context/Domain/OrderFeatureContext.php:
```diff
@@ -111,4 +186,119 @@ public function generateOrderInvoice($orderReference)
+     * @throws OrderException
+     * @throws RuntimeException
+     */
+    public function iUpdateOrdersWithIdsStatusTo(string $orderIdsString, string $status)
```
Just a suggestion, but nothing important: it's better to have "classical" function names for behat steps
```suggestion
    public function updateOrdersToStatuses(string $orderIdsString, string $status)
```

because steps can call other steps, so you can create macro steps:

```
    public function iCreateAFullOrder()
    {
        $this->createCustomer();
        $this->createAddress();
        ...
    }
```

matks commented on tests/Integration/Behaviour/Features/Scenario/Order/duplicate_order_cart.feature:
```diff
@@ -0,0 +1,18 @@
+  Background:
+    Given email sending is disabled
+    And the current currency is "USD"
+    And there is existing order with id 1
```
Maybe we dont need to check the very existence as the next step will check it anyway ?

We try to make behat tests as close as possible to real world usecases. It's not easy because our tests are most of the time related to CRUD operations, but the idea is to say "A user could describe this scenario as a feature to implement into a Ecommerce CMS".

So
`As a User, when I delete order "ABC123" then order "ABC123" has status "deleted"`
quite works while
`As a User, when I delete order with ID 123 then order 123 has status 5`
is more technical-oriented

It's very hard to do in our case, but we can try to walk this way by avoiding some mathematical checks such as "object X exists" and rather checks "object X name is 'red mug'" 😉

tdavidsonas88 replied:
@matks I totally agree to this and would like to do it more like user behavior when it comes to Behat, just the way I understand it getting orders by reference "ABC123" would be better but then it goes out of the Domain scope - uses legacy Object Model class as the current GetOrderForViewing is not capable to get order by reference and expects id (int type).

It would be really nice to find user oriented way to test code components with Behat.

For the current state of the Domain code written I see several ways to do that:
1. Test code components with technical oriented style but also write user behavior tests which interact with the UI - as Behat is designed for user behavior. OR
2. Improve our Domain code components to be able to use references like "ABC123.

Would be nice if we could find nice solution to this :)

And yes, we don't need to check the very existence

matks commented on tests/Integration/Behaviour/Features/Context/Domain/OrderPaymentFeatureContext.php:
```diff
@@ -0,0 +1,214 @@
+class OrderPaymentFeatureContext extends AbstractDomainFeatureContext
+{
+    /**
+     * @When I add payment to order with id :orderId with the following properties:
```
```suggestion
     * @When I add payment to order with id :orderId with the following details:
```
Following the trend "try to make it as human readable as possible" could we try to make this operations more like "And I pay order X using a check whose amount is 5.548"

I suggest it for this step because this step is actually a very human behavior: paying for an online order is something that happens in a classical ecommerce checkout process 😄

matks commented on tests/Integration/Behaviour/Features/Context/Domain/OrderPaymentFeatureContext.php:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    /**
+     * @Then if I query order with id :orderId payments I should get :numberOfPayments payments
```
```suggestion
     * @Then order :orderId should have :numberOfPayments payments
```
Make it short, make it clear, make it readable 😉

matks commented on tests/Integration/Behaviour/Features/Context/Domain/OrderPaymentFeatureContext.php:
```diff
@@ -0,0 +1,214 @@
+     * @throws ServiceCircularReferenceException
+     * @throws ServiceNotFoundException
+     */
+    public function ifIQueryOrderWithIdPaymentsIShouldGetAnOrderWithProperties(int $orderId, TableNode $table)
```
Can you try to use `DataComparator` functions, like in https://github.com/PrestaShop/PrestaShop/blob/develop/tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php#L197 ?

And if it does not do the job ... do you think you can try to improve it ? The kind of complex comparisons you are doing here, we can guess we'll have to do many more, so let's try to have something reusable to do it for us 😄

tdavidsonas88 replied:
For start will create OrderPaymentForViewing object from table properties and use PHPUnit comparator (assertEquals) to compare those two objects. Let's see the power of PHPUnit object comparator :) The creation of object will be reusable.

That's the quickest solution. Might not though the best.

matks commented:
OK @tdavidsonas88 I am sorry for the late review, here it is

1) Very nice job, thank you for doing it although it can be quite boring, especially writing tests AFTER the code
2) I have some small feedbacks but the PR is mainly good, I could even merge it like it is now 😄 
3) The feedbacks I provided are mainly about having scenarios as human-readable and customer-oriented as possible, and about code reusability

tdavidsonas88 committed:
fixes after code review

tdavidsonas88 commented:
@matks Thank you for detailed review with explanations. It was not boring because it was interesting to figure out business value of each feature for Behat tests and see the business logic implemented. 

I have fixed most of the code related to your feedback already. Also left some feedback about the way PrestaShop uses Behat. It is not 100% clear for me why this approach was chosen (testing components instead of user behavior UI included) so sometimes I have inner conflicts in my mind about that. But hope to understand it more clearly soon.

Thank you very much for your feedback :) it was interesting to resolve the code according them.

tdavidsonas88 committed:
work in progress commit

tdavidsonas88 committed:
Quality improvements after code review - will be more soon

tdavidsonas88 committed:
Refactored orders tests after code review recommendations

tdavidsonas88 committed:
Refactored orders tests after code review recommendations

tdavidsonas88 committed:
Reduced legacy classes usage in the Behat tests;

tdavidsonas88 committed:
Feature: Order from Back Office (BO) code optimized

tdavidsonas88 committed:
SearchProductsHandler test coverage

tdavidsonas88 committed:
SearchProductsHandler test coverage

tdavidsonas88 committed:
written todo's for tests not to depend on the legacy classes

tdavidsonas88 committed:
written todo's for tests not to depend on the legacy classes

tdavidsonas88 committed:
ChangeOrderDeliveryAddressHandler test

tdavidsonas88 committed:
ChangeOrderDeliveryAddressCommand test covered

tdavidsonas88 committed:
ChangeOrderDeliveryAddressCommand test covered

tdavidsonas88 committed:
more specific todo's

tdavidsonas88 committed:
todos left, addresses handlers still depends on legacy address details

tdavidsonas88 committed:
todos suggestions to be implemented

matks commented:
> I have fixed most of the code related to your feedback already. Also left some feedback about the way PrestaShop uses Behat. It is not 100% clear for me why this approach was chosen (testing components instead of user behavior UI included) so sometimes I have inner conflicts in my mind about that. But hope to understand it more clearly soon.

For a long time, PrestaShop was provided with a small test coverage using PHPUnit.
PHPUnit powered unit tests (that was OK) and components tests (using database). These component tests were badly written and implemented and we had multiple issues where tests would fail although the code was fine.

We decided to put these tests into tests-legacy folder*.

PrestaShop has also been provided since a few years with UI tests, powered first by Selenium then by Puppeeter. These tests check the full stack (= they run the PrestaShop application and interact with it from a browser controlled by code). These tests are written by the QA team.

So Behat was introduced in order to replace old badly written components tests. It could have been used for UI testing as well, but QA team has already a full stack to deal with it so we'll stick to component testing with behat, at least for now.

*We did not delete them because they have some values, but we want to replace them with better tests.

matks commented:
@tdavidsonas88 Seems OK to me now 👍 

Can you tell me why the "WIP" label is still in the PR's name ? Are you still working on this PR ?

tdavidsonas88 committed:
AddressFeatureContext created

tdavidsonas88 committed:
Scenario: Change order shipping address

tdavidsonas88 commented:
> @tdavidsonas88 Seems OK to me now 
> 
> Can you tell me why the "WIP" label is still in the PR's name ? Are you still working on this PR ?

Thank you for useful comments @matks Still there are some Handlers not covered with tests related to Orders and I do changes to this branch. However merge it when you can and I will create new branch from develop to cover those changes. This branch is already big one.

tdavidsonas88 committed:
code style fixes

tdavidsonas88 committed:
code style fixes

tdavidsonas88 committed:
order feature small changes

tdavidsonas88 commented:
@matks Finished working on this branch; It can be merged.

matks commented:
Closing and reopening in order to trigger Travis again as it seems payload for CI was lost

matks commented:
@tdavidsonas88 I'm sorry we have a behat test failing 😐 

```
Scenario: pay order with negative amount and see it is not valid  # Features/Scenario/Order/order_from_bo.feature:47

    When order "bo_order1" has 0 payments                           # Tests\Integration\Behaviour\Features\Context\Domain\OrderPaymentFeatureContext::getOrderPayments()
    And I pay order "bo_order1" with the invalid following details: # Tests\Integration\Behaviour\Features\Context\Domain\OrderPaymentFeatureContext::addPaymentToOrderWithTheInvalidFollowingProperties()
      | date           | 2019-11-26 13:56:22 |
      | payment_method | Payments by check   |
      | transaction_id | test!@#$%%^^&* OR 1 |
      | id_currency    | 1                   |
      | amount         | -5.548              |
    Then order "bo_order1" has 0 payments                           # Tests\Integration\Behaviour\Features\Context\Domain\OrderPaymentFeatureContext::getOrderPayments()
      Order "bo_order1" number of payments  is "1", but "0" was expected (RuntimeException)
```

It would look like a badly initialized data 🤔 where order "bo_order1" already has 1 payment

tdavidsonas88 commented:
> Scenario: pay order with negative amount and see it is not valid

@matks this test failing is related to this issue: https://github.com/PrestaShop/PrestaShop/issues/16582
It is not an issue, let me know and will fix the test. 

Adding payment with negative amount gives total sum negative in this case and an unexpected error is thrown to the Front Office.

matks commented:
@tdavidsonas88 I disabled (= added comment statements) but another test prevents the merge:

```

  Scenario: Change order shipping address                                           # Features/Scenario/Order/order_from_bo.feature:132
    Given I create new address with following details:                              # Tests\Integration\Behaviour\Features\Context\Domain\AddressFeatureContext::createNewAddressWithFollowingDetails()
      | Customer email  | pub@prestashop.com   |
      | Address alias   | dummyCustomerAddress |
      | First Name      | CustomerName         |
      | Last Name       | CustomerSurname      |
      | Address         | Street st. 1         |
      | Zip/Postal Code | 11111                |
      | City            | Paris                |
      | Country         | France               |
      TODO: write pending definition
```

Can you handle this one so we can merge it ?

tdavidsonas88 committed:
removed todos

tdavidsonas88 committed:
fixes for change order shipping test case

tdavidsonas88 committed:
pending definition removed

matks commented:
Thank you @tdavidsonas88
